core/application/reconcile.go

```diff
@@ -52,52 +57,99 @@ func (a *app) ResyncMicroVMs(ctx context.Context, namespace string) error {
 	return nil
 }
 
-func (a *app) plan(spec *models.MicroVM, logger *logrus.Entry) (planner.Plan, error) {
+func (a *app) plan(spec *models.MicroVM, logger *logrus.Entry) planner.Plan {
 	l := logger.WithField("stage", "plan")
 	l.Info("Generate plan")
 
-	if spec.Status.Retry > a.cfg.MaximumRetry {
-		return nil, reachedMaximumRetryError{vmid: spec.ID, retries: spec.Status.Retry}
-	}
-
 	// Delete only if the spec was marked as deleted.
 	if spec.Spec.DeletedAt != 0 {
 		input := &plans.DeletePlanInput{
 			StateDirectory: a.cfg.RootStateDir,
 			VM:             spec,
 		}
 
-		return plans.MicroVMDeletePlan(input), nil
+		return plans.MicroVMDeletePlan(input)
 	}
 
 	input := &plans.CreateOrUpdatePlanInput{
 		StateDirectory: a.cfg.RootStateDir,
 		VM:             spec,
 	}
 
-	return plans.MicroVMCreateOrUpdatePlan(input), nil
+	return plans.MicroVMCreateOrUpdatePlan(input)
+}
+
+func (a *app) reschedule(ctx context.Context, logger *logrus.Entry, spec *models.MicroVM) error {
+	spec.Status.Retry++
+	waitTime := time.Duration(spec.Status.Retry*backoffBaseInSeconds) * time.Second
+	spec.Status.NotBefore = time.Now().Add(waitTime).Unix()
+
+	logger.Infof(
+		"[%d/%d] reconciliation failed, reschedule at %s",
+		spec.Status.Retry,
+		a.cfg.MaximumRetry,
+		time.Unix(spec.Status.NotBefore, 0),
+	)
+
+	if _, err := a.ports.Repo.Save(ctx, spec); err != nil {
+		return fmt.Errorf("saving spec after plan failed: %w", err)
+	}
+
+	go func(id, ns string, sleepTime time.Duration) {
+		logger.Info("Wait to emit update")
+		time.Sleep(sleepTime)
+		logger.Info("Emit pdate")
+
+		_ = a.ports.EventService.Publish(
+			context.Background(),
+			defaults.TopicMicroVMEvents,
+			&events.MicroVMSpecUpdated{
+				ID:        id,
+				Namespace: ns,
+			},
+		)
+	}(spec.ID.Name(), spec.ID.Namespace(), waitTime)
+
+	return nil
 }
 
 func (a *app) reconcile(ctx context.Context, spec *models.MicroVM, logger *logrus.Entry) error {
-	l := logger.WithField("vmid", spec.ID.String())
-	l.Info("Starting reconciliation")
+	localLogger := logger.WithField("vmid", spec.ID.String())
+	localLogger.Info("Starting reconciliation")
 
-	plan, planErr := a.plan(spec, l)
-	if planErr != nil {
-		return planErr
+	if spec.Status.Retry > a.cfg.MaximumRetry {
+		spec.Status.State = models.FailedState
+
+		logger.Error(reachedMaximumRetryError{vmid: spec.ID, retries: spec.Status.Retry})
+
+		return nil
+	}
+
+	if spec.Status.NotBefore > 0 && time.Now().Before(time.Unix(spec.Status.NotBefore, 0)) {
+		return nil
 	}
 
+	plan := a.plan(spec, localLogger)
+
 	execCtx := portsctx.WithPorts(ctx, a.ports)
 
 	executionID, err := a.ports.IdentifierService.GenerateRandom()
 	if err != nil {
+		if scheduleErr := a.reschedule(ctx, localLogger, spec); scheduleErr != nil {
```

Comment:
If we can't get a random number should we re-schedule? Seems like that wouldn't be a transient error we should fail straight away.

Reply:
I don't understand this. What is the random number?

Reply:
Ok I see :rofl:

Reply:
I'm not sure, even if it seems a transient error, if it's a problem on the host, all vms will reconcile in an infinite loop. Reschedule prevent this with increasing the Retry value by one. If it's a transient error, the next execution will happen in 20s, but we don't get an infinite loop in case something is wrong with the host.

Reply:
My point is that i don't see this being a transient error in this situation, and so retrying up to the maximum number of retries won't change anything. Can we fail fast and stop reconciliation?

Reply:
So like terminate flintlock if this happens?

Reply:
It feels like it shouldn't continue to run as i don't see a situation where it would start working again. Can you see this ever being a transient error which will be corrected on retry?

And maybe we try generating a random number early using the service before we start the reconciler? wdyt?

Reply:
Well, reading ulid, I think we can get error if the time is too big or the entropy increment hits overflow. I don't think any of these can be resolved without restarting the application or even the server (not the resync). The rest is when it was not able to read from bytes that can happen is the system is very busy and can't read from urandom because of some kind of io error, that can be resolved over time.

Reply:
Thats great to understand the scenarios where it may fail, thanks for looking at this....great :+1: